File: matiasmasso/Mat.NET/DTO/DTOProductBrand.cs
using MatHelperStd;
using Newtonsoft.Json;

using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOProductBrand : DTOProduct
    {

        public class Treenode
        {
            public Guid Guid { get; set; }
            public DTOLangText.Compact Nom { get; set; }
            public List<DTOProductCategory.Treenode> Categories { get; set; }

            public static Treenode Factory(Guid guid)
            {
                Treenode retval = new Treenode();
                retval.Guid = guid;
                retval.Categories = new List<DTOProductCategory.Treenode>();
                return retval;
            }

            public override string ToString()
            {
                string retval = "Brand.Treenode Guid:" + this.Guid.ToString();
                if (this.Nom != null)
                    retval += " " + this.Nom.Esp;
                return retval;
            }

            public DTOProductBrand ToProductBrand()
            {
                DTOProductBrand retval = new DTOProductBrand(this.Guid);
                retval.Nom = this.Nom.toLangText();
                return retval;
            }

        }

        public const int LogoWidth = 150;
        public const int LogoHeight = 48;

        public DTOEmp Emp { get; set; }
        public int Id { get; set; }
        public DTOProveidor Proveidor { get; set; }
        public CodDists CodDist { get; set; }
        [JsonIgnore]
        public Byte[] Logo { get; set; }
        [JsonIgnore]
        public Byte[] LogoDistribuidorOficial { get; set; }
        public bool ShowAtlas { get; set; }
        public bool EnabledxConsumer { get; set; }
        public bool EnabledxPro { get; set; }

        public DateTime FchLastEdited { get; set; } // per sitemap.xml

        public List<DTOProductCategory> Categories { get; set; }
        public DTODept.Collection Depts { get; set; }

        public DTOCliProductBlocked.Codis CliProductBlockedCodi { get; set; }

        public DTOCnap Cnap { get; set; }

        public DTOCountry MadeIn { get; set; }
        public new DTOCodiMercancia CodiMercancia { get; set; }

        public int WebAtlasDeadline { get; set; } // Max Days from last order to be published as distributor
        public int WebAtlasRafflesDeadline { get; set; } // Max Days from last order to be published as raffle sale point
        public bool EnLiquidacio { get; set; }

        public enum CodDists
        {
            Free,
            DistribuidorsOficials
        }

        public enum CodStks
        {
            NotSet,
            Intern // nosaltres estoquem. Entra quan comprem, surt quan venem
    ,
            Extern // ells estoquen; entra automáticament tot allo i nomes allo que surt
        }

        public enum Wellknowns
        {
            Britax,
            Romer,
            Bob,
            Inglesina,
            FourMoms,
            TommeeTippee,
            FisherPrice,
            JBimbi,
            Gro,
            Varios,
            None
        }

        public DTOProductBrand() : base()
        {
            base.SourceCod = SourceCods.Brand;
            this.Categories = new List<DTOProductCategory>();
            this.Depts = new DTODept.Collection();
        }

        public DTOProductBrand(Guid oGuid) : base(oGuid)
        {
            base.SourceCod = SourceCods.Brand;
            this.Categories = new List<DTOProductCategory>();
            this.Depts = new DTODept.Collection();
        }

        public static DTOProductBrand Wellknown(DTOProductBrand.Wellknowns id)
        {
            DTOProductBrand retval = null;
            string sGuid = "";
            switch (id)
            {
                case DTOProductBrand.Wellknowns.Inglesina:
                    {
                        sGuid = "B1A0FB03-0C18-4607-9091-DF5A6A635BB0";
                        break;
                    }

                case DTOProductBrand.Wellknowns.Britax:
                    {
                        sGuid = "D56CE172-3C98-48E0-A378-8718BE8622F7";
                        break;
                    }

                case DTOProductBrand.Wellknowns.Romer:
                    {
                        sGuid = "D4C2BC59-046D-42D3-86E3-BDCA91FB473F";
                        break;
                    }

                case DTOProductBrand.Wellknowns.Bob:
                    {
                        sGuid = "63F67FDB-812F-49F9-B06C-023EE8A984EC";
                        break;
                    }

                case DTOProductBrand.Wellknowns.FisherPrice:
                    {
                        sGuid = "7C097674-233E-4899-92A7-37F37DD6D1F4";
                        break;
                    }

                case DTOProductBrand.Wellknowns.FourMoms:
                    {
                        sGuid = "67058F90-1FD6-4AE6-82ED-78447779B358";
                        break;
                    }

                case DTOProductBrand.Wellknowns.TommeeTippee:
                    {
                        sGuid = "B55B006D-3322-4E41-8CF7-9A02C3503A09";
                        break;
                    }

                case Wellknowns.Gro:
                    {
                        sGuid = "7b5dd269-62ea-481e-8be3-6e4c9d772117";
                        break;
                    }

                case DTOProductBrand.Wellknowns.Varios:
                    {
                        sGuid = "93995799-2865-46E0-A0EC-05DB6A80C7B4";
                        break;
                    }
            }

            if (sGuid.isNotEmpty())
            {
                Guid oGuid = new Guid(sGuid);
                retval = new DTOProductBrand(oGuid);
            }
            return retval;
        }

        public Wellknowns wellknownId()
        {
            String sGuid = base.Guid.ToString();
            foreach (Wellknowns wk in Enum.GetValues(typeof(Wellknowns)))
            {
                if (Wellknown(wk).Guid.Equals(base.Guid))
                {
                    return wk;
                }
            }
            return Wellknowns.None;
        }
        public static string Aliases(DTOProductBrand oBrand)
        {
            if (oBrand.Equals(Wellknown(Wellknowns.Romer)))
            {
                return "britax, romer, roemer, römer, britax romer, britax roemer, britax-roemer, britax-romer, britax_romer, britax römer";
            }
            else if (oBrand.Equals(Wellknown(Wellknowns.TommeeTippee)))
            {
                return "tommee_tippee,tommee tippee,tommeetippee,tometipe,tomitipi,tome tipe,tomi tipi";
            }
            else if (oBrand.Equals(Wellknown(Wellknowns.FourMoms)))
            {
                return "4moms,fourmoms,formoms";
            }
            else if (oBrand.Equals(Wellknown(Wellknowns.JBimbi)))
            {
                return "jbimbi";
            }
            else if (oBrand.Equals(Wellknown(Wellknowns.FisherPrice)))
            {
                return "fisher-price,fisherprice,fisher price";
            }
            else
            {
                return "";
            }
        }

        public bool Matches(string searchTerm)
        {
            return Nom.Contains(searchTerm);
        }

        public static DTOProductBrand FromNom(string nom)
        {
            DTOProductBrand retval = null;
            var wellknownIds = Enum.GetValues(typeof(Wellknowns));
            foreach (Wellknowns wellknownId in wellknownIds)
            {
                DTOProductBrand oBrand = Wellknown(wellknownId);
                string brandAliases = wellknownId.ToString();
                if (oBrand != null)
                {
                    brandAliases = Aliases(oBrand);
                }
                List<string> aliasList = brandAliases.Split(',').ToList();
                List<string> cleanAliasList = aliasList.Where(x => x.isNotEmpty()).ToList();
                string lowerNom = nom.ToLowerInvariant();
                if (cleanAliasList.Any(x => x.Trim().ToLowerInvariant() == lowerNom))
                {
                    retval = oBrand;
                    break;
                }
            }
            return retval;
        }

        public new string ThumbnailUrl(bool absoluteUrl = false)
        {
            return MmoUrl.ApiUrl("productBrand/thumbnail", base.Guid.ToString()); //falta aplicar a WebApi com passa a categoria
        }


        public DTOProductBrand clon()
        {
            List<Exception> exs = new List<Exception>();
            DTOProductBrand retval = new DTOProductBrand();
            DTOBaseGuid.CopyPropertyValues<DTOProductBrand>(this, retval, exs);
            return retval;
        }


        public static DTOProductBrand Factory(DTOEmp oEmp)
        {
            DTOProductBrand retval = new DTOProductBrand();
            retval.Emp = oEmp;
            return retval;
        }

        public string getUrl(DTOProduct.Tabs oTab = DTOProduct.Tabs.general, bool AbsoluteUrl = false, DTOLang lang = null)
        {
            var retval = MmoUrl.Factory(AbsoluteUrl, DTOProductBrand.urlSegment(this));

            if (oTab != DTOProduct.Tabs.general)
                retval = retval + "/" + DTOProduct.TabUrlSegment(oTab).Tradueix(lang);

            return retval;
        }



        public static string urlSegment(DTOProductBrand oBrand)
        {
            string retval = "";
            if (oBrand != null)
                retval = oBrand.urlSegment();
            return retval;
        }

        public string urlSegment(DTOProduct.Tabs oTab = DTOProduct.Tabs.general, DTOLang lang = null)
        {
            string retval = "";
            if (lang == null) lang = DTOLang.Default();
            if (this.Equals(DTOProductBrand.Wellknown(Wellknowns.Romer)))
                retval = "britax-roemer";
            else
                retval = MatHelperStd.UrlHelper.EncodedUrlSegment(this.Nom.Esp);

            if (oTab != DTOProduct.Tabs.general)
                retval = string.Format("{0}/{1}", retval, DTOProduct.TabUrlSegment(oTab).Tradueix(lang));

            return retval;
        }


        public string UrlSegmentSalePointsPerArea(string AreaNom)
        {
            return string.Format("/{0}/{1}", urlSegment(), AreaNom);
        }

        public string LogoUrl()
        {
            return MmoUrl.ApiUrl("productBrand/logo", base.Guid.ToString());
        }

        public static DTOGuidNom ToGuidNom(DTOProductBrand oBrand)
        {
            DTOGuidNom retval = new DTOGuidNom(oBrand.Guid, oBrand.Nom.Esp);
            return retval;
        }

        public static List<DTOGuidNom> ToGuidNoms(IEnumerable<DTOProductBrand> oBrands)
        {
            List<DTOGuidNom> retval = new List<DTOGuidNom>();
            foreach (var oBrand in oBrands)
            {
                var item = DTOProductBrand.ToGuidNom(oBrand);
                retval.Add(item);
            }
            return retval;
        }

        public SpriteHelper.Sprite DeptsSprite(List<DTODept> oDepts, DTOLang lang)
        {
            var url = MmoUrl.ApiUrl("depts/sprite", base.Guid.ToString());
            var retval = SpriteHelper.Factory(url, DTODept.IMAGEWIDTH, DTODept.IMAGEHEIGHT);
            foreach (var oDept in oDepts)
                retval.addItem(oDept.Nom.Tradueix(lang), oDept.GetUrl(lang));
            return retval;
        }

        public DTOProductBrand Clone()
        {
            DTOProductBrand retval = new DTOProductBrand(Guid);
            retval.Nom = Nom;
            return retval;
        }

        public new class Collection : List<DTOProductBrand>
        {

        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOProductBrand.vb

Public Class DTOProductBrand
    Inherits DTOProduct

    Public Const LogoWidth = 150
    Public Const LogoHeight = 48

    Property emp As DTOEmp
    Property id As Integer
    Shadows Property proveidor As DTOProveidor
    Property codDist As CodDists
    <JsonIgnore> Property logo As Image
    <JsonIgnore> Property logoDistribuidorOficial As Image
    Property showAtlas As Boolean
    Property enabledxConsumer As Boolean
    Property enabledxPro As Boolean

    Property title_Esp As String
    Property title_Cat As String
    Property title_Eng As String

    Property tagline_Esp As String
    Property tagline_Cat As String
    Property tagline_Eng As String

    Property restrictAtlasToPremiumLine As DTOPremiumLine

    Property fchLastEdited As DateTime 'per sitemap.xml

    Property categories As List(Of DTOProductCategory)

    Property cliProductBlockedCodi As DTOCliProductBlocked.Codis

    Property cnap As DTOCnap

    Property madeIn As DTOCountry
    Shadows Property codiMercancia As DTOCodiMercancia

    Property webAtlasDeadline As Integer 'Max Days from last order to be published as distributor
    Property webAtlasRafflesDeadline As Integer 'Max Days from last order to be published as distributor

    Public Enum CodDists
        Free
        DistribuidorsOficials
    End Enum

    Public Enum CodStks
        NotSet
        Intern 'nosaltres estoquem. Entra quan comprem, surt quan venem
        Extern 'ells estoquen; entra automáticament tot allo i nomes allo que surt
    End Enum

    Public Enum wellknowns
        Britax
        Romer
        Bob
        Inglesina
        FourMoms
        TommeeTippee
        FisherPrice
        JBimbi
        Gro
        Varios
    End Enum

    Public Sub New()
        MyBase.New()
        MyBase.SourceCod = SourceCods.Brand
        _Categories = New List(Of DTOProductCategory)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        MyBase.SourceCod = SourceCods.Brand
        _Categories = New List(Of DTOProductCategory)
    End Sub

    Shared Function wellknown(id As DTOProductBrand.wellknowns) As DTOProductBrand
        Dim retval As DTOProductBrand = Nothing
        Dim sGuid As String = ""
        Select Case id
            Case DTOProductBrand.wellknowns.Inglesina
                sGuid = "B1A0FB03-0C18-4607-9091-DF5A6A635BB0"
            Case DTOProductBrand.wellknowns.Britax
                sGuid = "D56CE172-3C98-48E0-A378-8718BE8622F7"
            Case DTOProductBrand.wellknowns.Romer
                sGuid = "D4C2BC59-046D-42D3-86E3-BDCA91FB473F"
            Case DTOProductBrand.wellknowns.Bob
                sGuid = "63F67FDB-812F-49F9-B06C-023EE8A984EC"
            Case DTOProductBrand.wellknowns.FisherPrice
                sGuid = "7C097674-233E-4899-92A7-37F37DD6D1F4"
            Case DTOProductBrand.wellknowns.FourMoms
                sGuid = "67058F90-1FD6-4AE6-82ED-78447779B358"
            Case DTOProductBrand.wellknowns.TommeeTippee
                sGuid = "B55B006D-3322-4E41-8CF7-9A02C3503A09"
            Case wellknowns.Gro
                sGuid = "7b5dd269-62ea-481e-8be3-6e4c9d772117"
            Case DTOProductBrand.wellknowns.Varios
                sGuid = "93995799-2865-46E0-A0EC-05DB6A80C7B4"
        End Select

        If sGuid > "" Then
            Dim oGuid As New Guid(sGuid)
            retval = New DTOProductBrand(oGuid)
        End If
        Return retval
    End Function

    Public Function Clon() As DTOProductBrand
        Dim exs As New List(Of Exception)
        Dim retval As New DTOProductBrand
        DTOBaseGuid.CopyPropertyValues(Of DTOProductBrand)(Me, retval, exs)
        Return retval
    End Function


    Shared Shadows Function Factory(oEmp As DTOEmp) As DTOProductBrand
        Dim retval As New DTOProductBrand
        retval.Emp = oEmp
        Return retval
    End Function


    Public Function getUrl(Optional oTab As DTOProduct.Tabs = DTOProduct.Tabs.general, Optional AbsoluteUrl As Boolean = False) As String
        Dim retval = MmoUrl.Factory(AbsoluteUrl, DTOProductBrand.UrlSegment(Me))

        If oTab <> DTOProduct.Tabs.general Then
            retval = retval & "/" & oTab.ToString
        End If

        Return retval
    End Function

    Overloads Shared Function UrlSegment(oBrand As DTOProductBrand) As String
        Dim retval As String = ""
        If oBrand IsNot Nothing Then
            retval = oBrand.UrlSegment
        End If
        Return retval
    End Function

    Public Overloads Function UrlSegment(Optional oTab As DTOProduct.Tabs = DTOProduct.Tabs.general) As String
        Dim retval As String = ""
        If Me.Equals(DTOProductBrand.wellknown(wellknowns.Romer)) Then
            retval = "britax-roemer"
        Else
            retval = MatHelperStd.UrlHelper.EncodedUrlSegment(MyBase.Nom)
        End If

        If oTab <> DTOProduct.Tabs.general Then
            retval = String.Format("{0}/{1}", retval, oTab.ToString())
        End If

        Return retval
    End Function


    Public Function UrlSegmentSalePointsPerArea(AreaNom As String) As String
        Return String.Format("/{0}/{1}", UrlSegment(), AreaNom)
    End Function

    Shared Function ToGuidNom(oBrand As DTOProductBrand) As DTOGuidNom
        Dim retval As New DTOGuidNom(oBrand.Guid, oBrand.Nom)
        Return retval
    End Function

    Shared Function ToGuidNoms(oBrands As IEnumerable(Of DTOProductBrand)) As List(Of DTOGuidNom)
        Dim retval As New List(Of DTOGuidNom)
        For Each oBrand In oBrands
            Dim item = DTOProductBrand.ToGuidNom(oBrand)
            retval.Add(item)
        Next
        Return retval
    End Function

    Public Function DeptsSprite(oDepts As List(Of DTODept), oLang As DTOLang) As SpriteHelper.Sprite
        Dim url = MmoUrl.ApiUrl("depts/sprite", MyBase.Guid.ToString())
        Dim retval = SpriteHelper.Factory(url, DTODept.width, DTODept.height)
        For Each oDept In oDepts
            retval.addItem(oDept.LangNom.tradueix(oLang), oDept.url)
        Next
        Return retval
    End Function
End Class
